rfc/006-backfill-blocks.md

```diff
@@ -0,0 +1,155 @@
+the application with the following format:
+
+```golang
+type snapshot struct {
+ 	Height   uint64
+ 	Format   uint32
+ 	Chunks   uint32
+ 	Hash     []byte
+ 	Metadata []byte
+
+ 	trustedAppHash []byte // populated by light client
+}
+```
+
+The application will then be expected to parse information in the meta data
+regarding the backfill height and then return it in the response (alongside the
+app's decision):
+
+```proto
+message ResponseOfferSnapshot {
+  Result result = 1;
+  int64 backfill_height = 2;
+}
+```
+
+If all chunks are accepted and state sync is successful then the backfill height
+would be returned from state sync to the node so that it can execute the
+backfill process (via the blockchain reactor) to retrieve the necessary blocks.
+A backfill height of 0 would mean that this step is ignored.
+
+```golang
+go func() {
+		state, commit, backfill_height, err := ssR.Sync(stateProvider, config.DiscoveryTime)
+		if err != nil {
+			ssR.Logger.Error("State sync failed", "err", err)
+			return
+		}
+		err = stateStore.Bootstrap(state)
+		if err != nil {
+			ssR.Logger.Error("Failed to bootstrap node with new state", "err", err)
+			return
+		}
+		err = blockStore.SaveSeenCommit(state.LastBlockHeight, commit)
+		if err != nil {
+			ssR.Logger.Error("Failed to store last seen commit", "err", err)
+			return
+		}
+
+    // proposed method for backfilling blocks
+    bcR.BackfillBlocks(backfill_height)
+
+		if fastSync {
+			err = bcR.SwitchToFastSync(state)
+			if err != nil {
+				ssR.Logger.Error("Failed to switch to fast sync", "err", err)
+				return
+			}
+		} else {
+			conR.SwitchToConsensus(state, true)
+		}
+}
+```
+
+`BackfillBlocks` would be implemented similarly to fast sync itself, by
+requesting blocks and validating them by matching the hash of the new header
+to the LastBlockID hash in the trusted header. This would use the same block
+pool and thus would be exclusive with the fast sync process (only one can run
+at a time).
+
+For completeness, the node should also retrieve the `ValidatorSet`s,
+`ConsensusParam`'s and `ABCIResponse`'s for each of the heights. In alignment
+with the separation between state and block, it may makes sense that the state
+sync reactor rather than the blockchain reactor serve, request and verify this
+data.
+
+This could be done in one of two ways: use the light client via the
+RPC connection or creating a new channel to send this data through.
+
+However, this separation would require multiple reads to retrieve the header at
+each height and would make it difficult to coordinate with the process running
+on the blockchain reactor. Hence I would lean towards extending the blockchain
+reactor with new channels to request this data structure, to verify it and to
+persist it to the state store.
+
+The back filling of blocks on start up would occur synchronously, completing
+this action before switching to either fast sync or consensus.
```

Comment:
Blocking joining consensus is specific to validators? Can a full node do this asynchronously as it does not need to vote on blocks?

Reply:
Do full nodes try to validate blocks?

Reply:
In Tendermint-go they do, but do full nodes need to verify evidence? They can if they have the data available but it may not need to a requirement.

Reply:
Yeah see proposal no.2 I mention here: https://github.com/tendermint/tendermint/issues/5617#issue-736760581